### PR TITLE
feat(APIv2): RHICOMPL-3888 implement profiles controller

### PR DESCRIPTION
**Pull request by xmicha82**
## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [x] General Coding Practices

### CODE REVIEW

File: app/controllers/v2/profiles_controller.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    private
+
+    def profiles
+      @profiles ||= authorize(resolve_collection)
+    end
+
+    def profile
+      # rubocop:disable Rails/FindById
+      @profile ||= authorize(resource.find_by!(id: permitted_params[:id]))
+      # rubocop:enable Rails/FindById
```
Comment (skateman) [inline app/controllers/v2/profiles_controller.rb:25]: This is just temporary as the view we're using has no primary key.

Comment (skateman) [reply app/controllers/v2/profiles_controller.rb:25]: @xmicha82 please add a `# FIXME` above

File: app/models/v2/profile.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    self.table_name = 'canonical_profiles'
+
+    sortable_by :title
+
+    scoped_search on: :title, only_explicit: true, operators: %i[like unlike eq ne in notin]
+    scoped_search on: :ref_id, only_explicit: true, operators: %i[eq ne in notin]
+
+    belongs_to :security_guide, class_name: 'V2::SecurityGuide'
+
+    def self.count_by
```
Comment (skateman) [inline app/models/v2/profile.rb:15]: Same here. a `FIXME` is needed that after we swap tables this has to be removed.

File: spec/controllers/v2/concerns/sortable_spec.rb [hunk 3]
```diff
@@ -23,13 +23,21 @@
...
 
   let(:items) do
     tests[:entities].map do |entity|
-      entity = entity.dup
+      # entity = entity.each_with_object({}) do |(key, value), obj|
+      #   obj[key] = value
+      #   value.is_a?(String) && value.match(SUBSTITUTE_REGEX) do |m|
+      #     obj[key] = extra_params[m[1].to_sym]
+      #   end
+      # end
```
Comment (skateman) [inline spec/controllers/v2/concerns/sortable_spec.rb:39]: Why is this commented?

File: spec/rails_helper.rb [hunk 4]
```diff
@@ -23,7 +23,7 @@
 # directory. Alternatively, in the individual `*_spec.rb` files, manually
 # require only the support files necessary.
 #
-# Dir[Rails.root.join('spec/support/**/*.rb')].sort.each { |f| require f }
+# Dir[Rails.root.join('spec/controllers/v2/concerns/*.rb')].sort.each { |f| require f }
```
Comment (skateman) [inline spec/rails_helper.rb:26]: Why:?

File: spec/spec_helper.rb [hunk 5]
```diff
@@ -48,6 +48,17 @@ def nested_route(*parents)
   end
 end
 
+def factory_params(item, extra_params)
```
Comment (skateman) [inline spec/spec_helper.rb:51]: Maybe some explanation what this does?

File: spec/controllers/v2/concerns/searchable_spec.rb [hunk 6]
```diff
@@ -24,20 +24,29 @@
 #   :query: 'query to search by'
 # ```
 
-RSpec.shared_examples 'searchable' do
-  path = Rails.root.join('spec/fixtures/files/searchable', "#{described_class.name.demodulize.underscore}.yaml")
+RSpec.shared_examples 'searchable' do |*parents|
```
Comment (skateman) [inline spec/controllers/v2/concerns/searchable_spec.rb:27]: I guess this needs to be documented above, together with the `extra_params`

File: spec/controllers/v2/concerns/sortable_spec.rb [hunk 7]
```diff
@@ -23,13 +23,21 @@
 #  e.g.: expected order is 0, 1, 2, 3, but 2 and 3 need to be sorted by FactoryBot ID (sorted property is same for both)
 #     => written as [0, 1, [2, 3]]
 
-RSpec.shared_examples 'sortable' do
+RSpec.shared_examples 'sortable' do |*parents|
```
Comment (skateman) [inline spec/controllers/v2/concerns/sortable_spec.rb:26]: Same here, documentation please.

File: spec/controllers/v2/concerns/sortable_spec.rb [hunk 8]
```diff
@@ -22,14 +22,19 @@
 # `:result` field is an array of indexes into entities array with expected order, nested arrays will be sorted by id
 #  e.g.: expected order is 0, 1, 2, 3, but 2 and 3 need to be sorted by FactoryBot ID (sorted property is same for both)
 #     => written as [0, 1, [2, 3]]
+#
+# parents - array of parental ActiveRecord models
+# extra_params - extra parameters to be passed into request
```
Comment (skateman) [inline spec/controllers/v2/concerns/sortable_spec.rb:27]: This is not descriptive enough, where and how these would go? One of them is a `let` the other one is passed as an argument of the shared example. Also it is not explained what they actually do...

File: spec/controllers/v2/concerns/searchable_spec.rb [hunk 9]
```diff
@@ -23,21 +23,33 @@
...
 #   :query: 'query to search by'
 # ```
+#
+# parents - array of parental ActiveRecord models
+# extra_params - extra parameters to be passed into request
 
-RSpec.shared_examples 'searchable' do
-  path = Rails.root.join('spec/fixtures/files/searchable', "#{described_class.name.demodulize.underscore}.yaml")
+RSpec.shared_examples 'searchable' do |*parents|
+  SUBSTITUTE_REGEX = /^\$\{([a-zA-Z_]+)\}$/.freeze
```
Comment (skateman) [inline spec/controllers/v2/concerns/searchable_spec.rb:31]: Move this into the function...here it doesn't make any sense

File: spec/controllers/v2/concerns/metadata_spec.rb [hunk 10]
```diff
@@ -1,19 +1,25 @@
...
-RSpec.shared_examples 'with metadata' do
+RSpec.shared_examples 'with metadata' do |*parents|
   let(:meta_keys) { %w[total limit offset relationships] }
   let(:item_count) { 10 }
+  let(:extra_params) { {} }
 
   it 'contains total, limit, offset and relationships keys' do
-    get :index
+    nested_route(*parents) do |mocked_parents|
+      # binding.pry
```
Comment (romanblanco) [inline spec/controllers/v2/concerns/metadata_spec.rb:10]: left `binding.pry`

File: spec/controllers/v2/concerns/searchable_spec.rb [hunk 11]
```diff
@@ -23,21 +23,33 @@
...
       search[:entities][:not_found].map do |h|
-        FactoryBot.create(h.delete(:factory), **h)
+        entity = factory_params(h, extra_params)
+        FactoryBot.create(entity.delete(:factory), **entity)
       end
 
-      get :index, params: { filter: search[:query] }
+      nested_route(*parents) do |mocked_parents|
+        get :index, params: extra_params.merge(
+          { search: search[:query] },
```
Comment (skateman) [inline spec/controllers/v2/concerns/searchable_spec.rb:49]:
```suggestion
          { filter: search[:query] },
```

File: spec/spec_helper.rb [hunk 12]
```diff
@@ -52,6 +52,21 @@ def nested_route(*parents)
   end
 end
 
+# Assembles object to be passed into factory
+# - adds url params into an object based on yaml entity definition in sorting and searching specs
+SUBSTITUTE_REGEX = /^\$\{([a-zA-Z_]+)\}$/.freeze
```
Comment (skateman) [inline spec/spec_helper.rb:57]: Move this please into the method, it has no use elsewhere.

**Review by romanblanco — APPROVED** [review]
LGTM :+1:

**Pull request merged**